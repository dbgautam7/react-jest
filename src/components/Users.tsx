import { useState } from "react";

const Users = () => {
  const [name, setName] = useState("");
  const [btnName, setBtnName] = useState("");
  return (
    <div>
      <h1>Users___</h1>
      <p>This is the users list page 11.</p>
      <ul>
        <li>name 1</li>
        <li>name 2</li>
        <li>name 3</li>
      </ul>

      <input
        type="text"
        placeholder="Enter username"
        value={name}
        onChange={(e) => setName(e.target.value + " " + "Gautam")}
      />
      <br />
      <br />
      <br />
      <button onClick={() => setBtnName("Submit")}>Click Me</button>
      <p>{btnName}</p>
      <br />
      <br />
      <br />
      <img
        title="image from google"
        alt="image"
        src="https://images.pexels.com/photos/808465/pexels-photo-808465.jpeg?cs=srgb&dl=pexels-lastly-808465.jpg&fm=jpg"
      />
    </div>
  );
};

export default Users;
