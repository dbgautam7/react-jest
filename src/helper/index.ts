export const handleOtherMethodClick = () => {
  console.log("Hi");
  return "Hi";
};
