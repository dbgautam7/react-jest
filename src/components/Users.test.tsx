import { configure, fireEvent, render, screen } from "@testing-library/react";
import Users from "./Users";
import { handleOtherMethodClick } from "../helper";

configure({ testIdAttribute: "element-id" });

beforeEach(() => {
  console.log("Called before each test");
});

describe("Users", () => {
  test("renders heading", async () => {
    render(<Users />);
    expect(
      screen.getByRole("heading", { name: "Users___" })
    ).toBeInTheDocument();
  });

  test("renders paragraph", () => {
    render(<Users />);
    expect(
      screen.getByText(/This is the users list page/i)
    ).toBeInTheDocument();
  });

  test("renders a list of users", async () => {
    render(<Users />);
    const users = await screen.findAllByRole("listitem");
    expect(users).toHaveLength(3);
  });

  test("renders the correct user names", () => {
    render(<Users />);
    expect(screen.getByText("name 1")).toBeInTheDocument();
    expect(screen.getByText("name 2")).toBeInTheDocument();
    expect(screen.getByTitle("image from google")).toBeInTheDocument();
  });

  test("renders a <ul> element", () => {
    render(<Users />);
    expect(screen.getByRole("list")).toBeInTheDocument();
  });

  test("list items are inside the <ul>", () => {
    render(<Users />);
    const list = screen.getByRole("list");
    const items = screen.getAllByRole("listitem");
    items.forEach((item) => {
      expect(list).toContainElement(item);
    });
  });

  test("renders an input field", () => {
    render(<Users />);
    const input = screen.getByRole("textbox");
    expect(input).toBeInTheDocument();
    expect(input).toHaveAttribute("placeholder", "Enter username");
  });

  test.only("input field updates on change", () => {
    render(<Users />);
    const input: HTMLInputElement = screen.getByRole("textbox");
    fireEvent.change(input, { target: { value: "DB" } });
    expect((input as HTMLInputElement).value).toBe("DB Gautam");
    const inputValue = screen.getByDisplayValue("DB Gautam");
    expect(inputValue).toBeInTheDocument();
  });

  test("button click event test case", () => {
    render(<Users />);
    const btn = screen.getByRole("button");
    fireEvent.click(btn);
    expect(screen.getByText("Submit")).toBeInTheDocument();
  });

  test.only("button click event test case with data testid", () => {
    render(<Users />);
    const btn = screen.getByTestId("btn1");
    fireEvent.click(btn);
    expect(screen.getByText("Submit")).toBeInTheDocument();
  });

  test("button click event test case with other method in helper", () => {
    expect(handleOtherMethodClick()).toMatch("Hi");
    expect(handleOtherMethodClick()).toBe("Hi");
  });

  test("multiple button check with getAllByRole", () => {
    render(<Users />);
    const buttons = screen.getAllByRole("button");
    buttons.forEach((_, i) => {
      expect(buttons[i]).toBeInTheDocument();
    });
  });

  test("mutliple checkbox testcase using getAllByLabel", () => {
    render(<Users />);
    const checkboxes = screen.getAllByLabelText("Is Passed");
    for (let i = 0; i < checkboxes.length; i++) {
      expect(checkboxes[i]).toBeInTheDocument();
      expect(checkboxes[0]).toHaveClass("custom-checkbox");
      expect(checkboxes[1]).toHaveAttribute("id");
    }
  });
});

//snapshot test
// test("Snapshot for Sum", () => {
//   const container = render(<Users />);
//   expect(container).toMatchSnapshot();
// });

// afterEach(() => {
//   console.log("Called after each test");
// });

// We recommend installing an extension to run jest tests.
