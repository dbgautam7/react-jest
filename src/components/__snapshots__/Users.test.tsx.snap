// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot for Sum 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h1>
          Users___
        </h1>
        <p>
          This is the users list page 11.
        </p>
        <ul>
          <li>
            name 1
          </li>
          <li>
            name 2
          </li>
          <li>
            name 3
          </li>
        </ul>
        <input
          placeholder="Enter username"
          type="text"
          value=""
        />
        <br />
        <br />
        <br />
        <button>
          Click Me
        </button>
        <p />
        <br />
        <br />
        <br />
        <img
          alt="image"
          src="https://images.pexels.com/photos/808465/pexels-photo-808465.jpeg?cs=srgb&dl=pexels-lastly-808465.jpg&fm=jpg"
          title="image from google"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h1>
        Users___
      </h1>
      <p>
        This is the users list page 11.
      </p>
      <ul>
        <li>
          name 1
        </li>
        <li>
          name 2
        </li>
        <li>
          name 3
        </li>
      </ul>
      <input
        placeholder="Enter username"
        type="text"
        value=""
      />
      <br />
      <br />
      <br />
      <button>
        Click Me
      </button>
      <p />
      <br />
      <br />
      <br />
      <img
        alt="image"
        src="https://images.pexels.com/photos/808465/pexels-photo-808465.jpeg?cs=srgb&dl=pexels-lastly-808465.jpg&fm=jpg"
        title="image from google"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
